### ISSUE TITLE
func-names: Allow unnamed generators (option)

### ISSUE THREAD

**Issue by joegoldbeck** [eslint/eslint#9511]
**What rule do you want to change?**
func-names

**Does this change cause the rule to produce more or fewer warnings?**
Fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**
New option

**Please provide some example code that this change will affect:**

```js
const parse = coroutinify(function* (input) { ... })
```

**What does the rule currently do for this code?**
"Unexpected unnamed generator function" unless the option is set to `"never"`

**What will the rule do after it's changed?**
Allow unnamed generator functions if a second option is set to `{ generators: false }` or similar

**Comment by not-an-aardvark** [comment]
Thanks for the proposal. Could you clarify what the motivation for this option is? Why do you want to allow unnamed generators while disallowing other unnamed functions?

**Comment by joegoldbeck** [comment]
Sure, happy to clarify!

The reason comes down to language-level limitations, and in particular, that arrow notation can be used to create functions but not generators. 

When we create lightweight _functions_, we use arrow notation, which are inherently unnamed. As one reference point, the AirBnb styleguide [recommends this approach](https://github.com/airbnb/javascript#arrow-functions).

However, for _generators_, we can only use the standard `function*`, and so an exception for generators would allow us to create lightweight unnamed generators while leaving our stricter linting rules for functions intact.

**Comment by not-an-aardvark** [comment]
I see, thanks. That makes sense to me -- let's see if the rest of the team would be in favor of adding an option like this.

**Comment by ekimlinger** [comment]
Not sure if this would have to be a new rule request, however generators also spark a similar break with the 'func-style' rule. 

Say for example we have the following code:
``` javascript
const anExampleGenFunc= function* () {
    yield () => console.log('Do Something more than a console.log here');
};
```
I feel as though that this is perfectly valid JS (please correct me if I am wrong), but it raises a warning in eslint. 
`[eslint] Expected a function declaration. (func-style)`

**Comment by ilyavolodin** [comment]
It looks like there's more then enough support from the team to accept this. We just need somebody to champion this change to mark it as accepted. @eslint/eslint-team anyone willing to champion?

**Comment by not-an-aardvark** [comment]
@eslint/eslint-team Anyone willing to champion this change? If not, it's probably time to close it.

**Comment by platinumazure** [comment]
I'll champion this.

**Comment by OscarBarrett** [comment]
I'll submit a PR for this in the next couple of days.
